crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2539,17 +2539,21 @@ impl<'db> TypeInferenceBuilder<'db> {
     fn infer_arguments<'a>(
         &mut self,
         arguments: &'a ast::Arguments,
-        infer_as_type_expressions: bool,
+        infer_as_type_expressions: u32,
     ) -> CallArguments<'a, 'db> {
-        let infer_argument_type = if infer_as_type_expressions {
-            Self::infer_type_expression
-        } else {
-            Self::infer_expression
-        };
-
         arguments
             .arguments_source_order()
-            .map(|arg_or_keyword| {
+            .enumerate()
+            .map(|(index, arg_or_keyword)| {
+                // TODO: Remove this once we have proper overload matching
```

Comment:
I don't think overload matching is the feature that will allow removing this; `TypeForm` and contextual type inference might. But since the current approach is not buggy, just hardcoded, I don't think we need a TODO for that. We could make a note of it, but I'd probably put that in the doc comment for `takes_type_expression_arguments`, not here.
```suggestion
```